File: admin-web/tests/daire-tipi-rengi.test.ts
// (P122) TIP RENGI — panel ile mobilin AYNI rengi uretmesi kilidi.
//
// NEDEN KILIT: ayni daire tipi iki yuzeyde AYNI rengi almalidir. Yonetici
// panelde bakip mobilde dogruluyor; renk ayrisirsa iki ekrandan biri
// "yanlis" gorunur ve guven kaybi teknik bir hatadan pahaliya patlar.
// Iki ayri dilde yazilmis iki fonksiyonun ayni kalmasini ancak PAYLASILAN
// bir beklenen-deger tablosu garanti eder: asagidaki tablonun AYNISI
// `mobile/test/daire_tipi_rengi_test.dart` icinde de vardir. Biri
// degistirilirse o taraf duser.
import { describe, expect, it } from "vitest";

import {
  DAIRE_TIPI_PALETI,
  daireTipiKisa,
  daireTipiRengi,
} from "@/lib/daire-tipi-rengi";

/** PAYLASILAN TABLO — mobil testiyle BIREBIR ayni. */
const BEKLENEN: Record<string, string> = {
  "2+1": "#3949AB",
  "3+1": "#00897B",
  "1+1": "#5D4037",
  "4+1": "#8E24AA",
  Dubleks: "#8E24AA",
  Dükkan: "#00838F",
  Villa: "#C62828",
  Stüdyo: "#5D4037",
  "Bahçe Katı": "#43A047",
  "Çatı Katı": "#43A047",
};

describe("daireTipiRengi", () => {
  it("PAYLASILAN TABLOYU birebir uretir (mobil ile ayni)", () => {
    for (const [ad, renk] of Object.entries(BEKLENEN)) {
      expect(daireTipiRengi(ad), `tip: ${ad}`).toBe(renk);
    }
  });

  it("BOS/tanimsiz -> varsayilan indigo", () => {
    expect(daireTipiRengi(null)).toBe(DAIRE_TIPI_PALETI[0]);
    expect(daireTipiRengi(undefined)).toBe(DAIRE_TIPI_PALETI[0]);
    expect(daireTipiRengi("   ")).toBe(DAIRE_TIPI_PALETI[0]);
  });

  it("BUYUK/kucuk harf ve bosluk FARK ETMEZ", () => {
    expect(daireTipiRengi(" 2+1 ")).toBe(daireTipiRengi("2+1"));
    expect(daireTipiRengi("DUBLEKS")).toBe(daireTipiRengi("dubleks"));
  });

  it("KARARLI: ayni ad her cagride ayni rengi verir", () => {
    const bir = daireTipiRengi("Bahçe Katı");
    for (let i = 0; i < 50; i++) expect(daireTipiRengi("Bahçe Katı")).toBe(bir);
  });

  it("palet DISINA cikmaz", () => {
    for (const ad of ["a", "bb", "ccc", "çç", "🏠 Daire", "x".repeat(200)]) {
      expect(DAIRE_TIPI_PALETI).toContain(daireTipiRengi(ad));
    }
  });
});

describe("daireTipiKisa", () => {
  it("KISA ad oldugu gibi kalir", () => {
    expect(daireTipiKisa("2+1")).toBe("2+1");
    expect(daireTipiKisa("Dubleks")).toBe("Dubleks");
  });

  it("UZUN ad kirpilir; SONUC sinir kadar uzun olur", () => {
    // Sinir SONUCUN uzunlugudur (nokta dahil), kirpma noktasi degil:
    // 7 sinirinda "Dublek" + "…" = 7 karakter. Ilk yazimda beklenti
    // "Dubleks…" (8) idi ve test dustu — hucre genisligi hesabi sonucun
    // uzunluguna gore yapildigi icin dogru olan kodun davranisi.
    expect(daireTipiKisa("Dubleks Bahçe Katı")).toBe("Dublek…");
    expect([...daireTipiKisa("Dubleks Bahçe Katı")]).toHaveLength(7);
  });

  it("EMOJI/birlesik karakter ORTADAN BOLUNMEZ", () => {
    // `slice` kod BIRIMIYLE calissaydi vekil cifti yarilanir ve ekranda
    // bozuk karakter cikardi.
    const k = daireTipiKisa("🏠🏠🏠🏠🏠🏠🏠🏠🏠");
    expect([...k]).toHaveLength(7);
    expect(k.endsWith("…")).toBe(true);
  });

  it("BOS ad bos doner", () => {
    expect(daireTipiKisa(null)).toBe("");
    expect(daireTipiKisa("  ")).toBe("");
  });
});

File: mobile/test/daire_tipi_rengi_test.dart
/// (P122) TİP RENGİ — mobil ile panelin AYNI rengi üretmesi kilidi.
///
/// NEDEN KİLİT: aynı daire tipi iki yüzeyde **aynı** rengi almalıdır.
/// Yönetici panelde bakıp mobilde doğruluyor; renk ayrışırsa iki ekrandan
/// biri "yanlış" görünür ve güven kaybı teknik bir hatadan pahalıya patlar.
/// İki ayrı dilde yazılmış iki fonksiyonun aynı kalmasını ancak PAYLAŞILAN
/// bir beklenen-değer tablosu garanti eder: aşağıdaki tablonun AYNISI
/// `admin-web/tests/daire-tipi-rengi.test.ts` içinde de var. Biri
/// değiştirilirse o taraf düşer.
library;

import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/src/features/building_map/presentation/daire_tipi_rengi.dart';

/// PAYLAŞILAN TABLO — panel testiyle BİREBİR aynı.
const _beklenen = <String, int>{
  '2+1': 0xFF3949AB,
  '3+1': 0xFF00897B,
  '1+1': 0xFF5D4037,
  '4+1': 0xFF8E24AA,
  'Dubleks': 0xFF8E24AA,
  'Dükkan': 0xFF00838F,
  'Villa': 0xFFC62828,
  'Stüdyo': 0xFF5D4037,
  'Bahçe Katı': 0xFF43A047,
  'Çatı Katı': 0xFF43A047,
};

void main() {
  group('daireTipiRengi', () {
    test('PAYLASILAN TABLOYU birebir uretir (panel ile ayni)', () {
      _beklenen.forEach((ad, deger) {
        expect(daireTipiRengi(ad).toARGB32(), deger, reason: 'tip: $ad');
      });
    });

    test('BOS/null -> varsayilan indigo', () {
      expect(daireTipiRengi(null), daireTipiPaleti.first);
      expect(daireTipiRengi(''), daireTipiPaleti.first);
      expect(daireTipiRengi('   '), daireTipiPaleti.first);
    });

    test('BUYUK/kucuk harf ve bosluk FARK ETMEZ', () {
      // Yonetici tipi " 2+1 " diye girse de ayni rengi gormeli.
      expect(daireTipiRengi(' 2+1 '), daireTipiRengi('2+1'));
      expect(daireTipiRengi('DUBLEKS'), daireTipiRengi('dubleks'));
    });

    test('KARARLI: ayni ad her cagride ayni rengi verir', () {
      final bir = daireTipiRengi('Bahçe Katı');
      for (var i = 0; i < 50; i++) {
        expect(daireTipiRengi('Bahçe Katı'), bir);
      }
    });

    test('palet DISINA cikmaz', () {
      for (final ad in ['a', 'bb', 'ccc', 'çç', '🏠 Daire', 'x' * 200]) {
        expect(daireTipiPaleti, contains(daireTipiRengi(ad)));
      }
    });
  });

  group('daireTipiKisa', () {
    test('KISA ad oldugu gibi kalir', () {
      expect(daireTipiKisa('2+1'), '2+1');
      expect(daireTipiKisa('Dubleks'), 'Dubleks');
    });

    test('UZUN ad kirpilir; SONUC sinir kadar uzun olur', () {
      // Sinir SONUCUN uzunlugudur (nokta dahil): 7'de "Dublek" + "…".
      expect(daireTipiKisa('Dubleks Bahçe Katı'), 'Dublek…');
      expect(daireTipiKisa('Dubleks Bahçe Katı').length, 7);
    });

    test('BOS ad bos doner', () {
      expect(daireTipiKisa(null), '');
      expect(daireTipiKisa('  '), '');
    });
  });
}
